Vesper-Agents-Suite/xy-loss-agent/src/agent.ts

```diff
@@ -0,0 +1,130 @@
+import Web3 from "web3";
+import LRU from "lru-cache";
+
+import { TimeTracker, VesperFetcher } from 'vesper-forta-module';
+import { getJsonRpcUrl } from "forta-agent";
+import { BlockEvent, Finding, HandleBlock } from "forta-agent";
+import { Strategy_ABI, Borrow_Token_ABI } from "./abi";
+import poolAbi from "./pool.abi";
+import { createFinding } from "./utils";
+import BigNumber from "bignumber.js";
+const XY_STRATEGY_KEY = "_xyStrategies"
+const XY_STRATEGY_NAME = "XYStrategy"
+const web3: Web3 = new Web3(getJsonRpcUrl());
+const cache: LRU<string, string[]> = new LRU<string, string[]>({ max: 100 });
+const tracker: TimeTracker = new TimeTracker();
+const vesperFetcher: VesperFetcher = new VesperFetcher(web3);
+let cacheTime: number = 3600000; // one hour in milliseconds
+
+export const setCacheTime = (time: number) => {
+  cacheTime = time;
+}
+
+const fetchXYStrategies = async (
+  blockNumber: string
+): Promise<string[]> => {
+  if (blockNumber != "latest" && cache.get(blockNumber) !== undefined)
+    return cache.get(blockNumber) as string[];
+  let [V3] = await Promise.all([
+    vesperFetcher.getStrategiesV3(blockNumber)
+  ]);
+  V3 = Array.from(new Set<string>(V3));
+  const valueV3promise = V3.map(async (strat: string) => {
+    const sContract = new web3.eth.Contract(
+      Strategy_ABI,
+      strat,
+    );
+
+    const name: string = await sContract.methods.NAME().call({}, blockNumber);  
+    if (!name.includes(XY_STRATEGY_NAME)) return BigInt(0);
+
+    const pool: string = await sContract.methods.pool().call({}, blockNumber);
+    const borrowCTokenAddress: string = await sContract.methods.borrowCToken().call({}, blockNumber);
+    const borrowCToken = new web3.eth.Contract(
+      Borrow_Token_ABI,
+      borrowCTokenAddress,
+    );
+    const debtOfBorrowedToken = new BigNumber(await borrowCToken.methods.borrowBalanceStored(strat).call({}, blockNumber));
+    const poolContract = new web3.eth.Contract(poolAbi as any, pool);
+    const pricePerShare = new BigNumber(await poolContract.methods.pricePerShare().call({}, blockNumber));
+    const strategyBalance = new BigNumber(await poolContract.methods.balanceOf(strat).call({}, blockNumber));
+    const borrowedTokenAsset = pricePerShare.multipliedBy(strategyBalance).dividedBy(new BigNumber('1e18'))
```

Comment:
Do we know that the borrowedToken will always be an 18 decimal token?

Reply:
You are right, Need to get decimals from borrowedToken.
Also this needs some other update related to the pool. Will fix those as well.

Reply:
All pools are 18 decimals so we are good.
Detecting loss on borrow is sufficient . No need to include apy of x.